File: apps/mobile/lib/features/assistant/assistant_turn_request.dart
/// 小爱统一输入契约 TurnRequest（P2，对齐 Web `assistant_turn_request.ts`）。
library;

import 'assistant_scenes.dart';

typedef TurnSurface = String;

abstract final class TurnSurfaces {
  static const halfSheet = 'half_sheet';
  static const assistant = 'assistant';
  static const homePrefill = 'home_prefill';
  static const prewarm = 'prewarm';
}

class TurnHistoryMessage {
  const TurnHistoryMessage({required this.role, required this.content});
  final String role;
  final String content;

  Map<String, dynamic> toJson() => {'role': role, 'content': content};
}

class TurnRequest {
  const TurnRequest({
    this.anchorRef,
    this.question = '',
    this.scene,
    this.mode,
    required this.surface,
    this.history = const [],
    this.readerContext,
    this.knowledgeBaseId,
    this.conversationId,
    this.clientCapabilities,
  });

  final String? anchorRef;
  final String question;
  final AssistantScene? scene;
  final String? mode;
  final String surface;
  final List<TurnHistoryMessage> history;
  final Map<String, dynamic>? readerContext;
  final String? knowledgeBaseId;
  final String? conversationId;
  final Map<String, dynamic>? clientCapabilities;
}

class ResolvedTurnRequest extends TurnRequest {
  const ResolvedTurnRequest({
    super.anchorRef,
    required super.question,
    required this.refForApi,
    required this.scene,
    required this.mode,
    required super.surface,
    super.history,
    super.readerContext,
    super.knowledgeBaseId,
    super.conversationId,
    super.clientCapabilities,
  });

  final String? refForApi;
  @override
  final AssistantScene scene;
  @override
  final String mode;
}

ResolvedTurnRequest resolveTurnRequest(TurnRequest req) {
  final resolved = resolveChatTurn(
    anchorRef: req.anchorRef,
    historyLength: req.history.length,
    explicitScene: req.scene,
    mode: req.mode,
  );
  final mode = req.mode ?? resolved.scene.mode;
  return ResolvedTurnRequest(
    anchorRef: req.anchorRef,
    question: req.question,
    refForApi: resolved.refForApi,
    scene: resolved.scene,
    mode: mode,
    surface: req.surface,
    history: req.history,
    readerContext: req.readerContext,
    knowledgeBaseId: req.knowledgeBaseId,
    conversationId: req.conversationId,
    clientCapabilities: req.clientCapabilities,
  );
}

Map<String, dynamic> toChatStreamBody(ResolvedTurnRequest req) {
  return {
    if (req.refForApi != null && req.refForApi!.isNotEmpty) 'ref': req.refForApi,
    'question': req.question,
    'mode': req.mode,
    'scene': req.scene.id,
    if (req.history.isNotEmpty)
      'history': req.history.map((h) => h.toJson()).toList(),
    'surface': req.surface,
    if (req.readerContext != null) 'reader_context': req.readerContext,
    if (req.knowledgeBaseId != null && req.knowledgeBaseId!.isNotEmpty)
      'knowledge_base_id': req.knowledgeBaseId,
    if (req.conversationId != null && req.conversationId!.isNotEmpty)
      'conversation_id': req.conversationId,
    'client_capabilities': req.clientCapabilities ??
        const {
          'schema_version': 3,
          'supports_section_stream': true,
        },
  };
}

ResolvedTurnRequest buildHalfSheetTurnRequest({
  required String ref,
  required String question,
  required AssistantScene scene,
  List<TurnHistoryMessage> history = const [],
  Map<String, dynamic>? readerContext,
  String? knowledgeBaseId,
  String? conversationId,
}) {
  return resolveTurnRequest(
    TurnRequest(
      anchorRef: ref,
      question: question,
      scene: scene,
      surface: TurnSurfaces.halfSheet,
      history: history,
      readerContext: readerContext,
      knowledgeBaseId: knowledgeBaseId,
      conversationId: conversationId,
    ),
  );
}

ResolvedTurnRequest buildAssistantTurnRequest({
  required String question,
  String? anchorRef,
  AssistantScene? scene,
  String? mode,
  List<TurnHistoryMessage> history = const [],
  Map<String, dynamic>? readerContext,
  String? knowledgeBaseId,
}) {
  return resolveTurnRequest(
    TurnRequest(
      anchorRef: anchorRef,
      question: question,
      scene: scene,
      mode: mode,
      surface: TurnSurfaces.assistant,
      history: history,
      readerContext: readerContext,
      knowledgeBaseId: knowledgeBaseId,
    );
  );
}

File: apps/web/lib/assistant_turn_request.ts
/**
 * 小爱统一输入契约 TurnRequest（P2）。
 * 半屏 / Tab / Chip / 预填 均经 resolveTurnRequest → toChatStreamBody。
 */
import type { ChatHistoryTurn, ChatReaderContext, ChatStreamBody } from '@/lib/api_core';
import {
  resolveChatTurn,
  SCENES,
  type AssistantScene,
} from '@/lib/assistant_scenes';

export type TurnSurface =
  | 'half_sheet'
  | 'assistant'
  | 'home_prefill'
  | 'prewarm';

export type ClientCapabilities = {
  schema_version?: number;
  supports_section_stream?: boolean;
};

export type TurnHistoryMessage = ChatHistoryTurn;

export type TurnRequest = {
  anchorRef?: string | null;
  selectionText?: string;
  question: string;
  displayQuestion?: string;
  scene?: AssistantScene | null;
  mode?: string;
  surface: TurnSurface;
  history?: TurnHistoryMessage[];
  readerContext?: ChatReaderContext;
  knowledgeBaseId?: string | null;
  conversationId?: string | null;
  clientCapabilities?: ClientCapabilities;
};

export type ResolvedTurnRequest = TurnRequest & {
  refForApi: string | null;
  scene: AssistantScene;
  mode: string;
};

/** 客户端 scene 建议 + ref 策略（服务端 meta 为权威）。 */
export function resolveTurnRequest(req: TurnRequest): ResolvedTurnRequest {
  const { refForApi, scene } = resolveChatTurn({
    anchorRef: req.anchorRef,
    historyLength: req.history?.length ?? 0,
    explicitScene: req.scene ?? undefined,
    mode: req.mode,
  });
  const mode = req.mode ?? SCENES[scene].mode;
  return { ...req, refForApi, scene, mode };
}

export function toChatStreamBody(req: ResolvedTurnRequest): ChatStreamBody {
  return {
    ref: req.refForApi ?? undefined,
    question: req.question,
    mode: req.mode,
    scene: req.scene,
    history: req.history,
    surface: req.surface,
    reader_context: req.readerContext,
    knowledge_base_id: req.knowledgeBaseId,
    conversation_id: req.conversationId ?? undefined,
    client_capabilities: req.clientCapabilities ?? {
      schema_version: 3,
      supports_section_stream: true,
    },
  };
}

/** 读经半屏一轮请求。 */
export function buildHalfSheetTurnRequest(opts: {
  ref: string;
  question: string;
  scene: AssistantScene;
  history?: TurnHistoryMessage[];
  readerContext?: ChatReaderContext;
  knowledgeBaseId?: string | null;
  conversationId?: string | null;
}): ResolvedTurnRequest {
  return resolveTurnRequest({
    anchorRef: opts.ref,
    question: opts.question,
    scene: opts.scene,
    surface: 'half_sheet',
    history: opts.history,
    readerContext: opts.readerContext,
    knowledgeBaseId: opts.knowledgeBaseId,
    conversationId: opts.conversationId,
  });
}

/** 小爱 Tab 一轮请求。 */
export function buildAssistantTurnRequest(opts: {
  question: string;
  anchorRef?: string | null;
  scene?: AssistantScene | null;
  mode?: string;
  history?: TurnHistoryMessage[];
  readerContext?: ChatReaderContext;
  knowledgeBaseId?: string | null;
  conversationId?: string | null;
}): ResolvedTurnRequest {
  return resolveTurnRequest({
    anchorRef: opts.anchorRef,
    question: opts.question,
    scene: opts.scene,
    mode: opts.mode,
    surface: 'assistant',
    history: opts.history,
    readerContext: opts.readerContext,
    knowledgeBaseId: opts.knowledgeBaseId,
    conversationId: opts.conversationId,
  });
}
